cas-server-core-authentication/src/main/java/org/jasig/cas/authentication/DefaultAuthenticationSupervisor.java

```diff
@@ -0,0 +1,110 @@
+package org.jasig.cas.authentication;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+/**
+ * This is {@link DefaultAuthenticationSupervisor}.
+ *
+ * @author Misagh Moayyed
+ * @since 4.2.0
+ */
+@Component("defaultAuthenticationSupervisor")
+public final class DefaultAuthenticationSupervisor implements AuthenticationSupervisor {
+    /**
+     * The constant LOGGER.
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultAuthenticationSupervisor.class);
+
+    /**
+     * The Authentication manager.
+     */
+    @Autowired
+    @Qualifier("authenticationManager")
+    private AuthenticationManager authenticationManager;
+
+    /**
+     * The Authentication context builder.
+     */
+    @Autowired
+    @Qualifier("defaultAuthenticationContextBuilder")
+    private AuthenticationContextBuilder authenticationContextBuilder;
+
+    /**
+     * Instantiates a new Default authentication supervisor.
+     */
+    public DefaultAuthenticationSupervisor() {}
+
+    @Override
+    public void authenticate(final Credential... credentials) throws AuthenticationException {
+        final Set<Credential> sanitizedCredentials = sanitizeCredentials(credentials);
+        if (!sanitizedCredentials.isEmpty()) {
+            final Credential[] sanitizedCredentialsArray = sanitizedCredentials.toArray(new Credential[] {});
+            LOGGER.debug("Sanitized credentials [{}] prior to authentication", sanitizedCredentialsArray);
+            final Authentication authentication =
+                    this.authenticationManager.authenticate(sanitizedCredentialsArray);
+            LOGGER.debug("Successful authentication; Collecting authentication result [{}]", authentication);
+            this.authenticationContextBuilder.collect(authentication);
```

Comment:
We attach the authentication to the TGT object, not the webflow. The builder produces one final authentication object that is a conglomerate. It contains everything about the authentication event, collected from all authentication attempts. That translates to a deployment, where the primary flow executes authN1, and secondary flows execute the remaining authN events. The final result produced by the builder is attached to the TGT when CASImpl generates the TGT. 

You may very well be right about the threading issue. I'll review and run some additional tests to see if I can move the builder to a local instance, or a thread-safe instance.